File: afamar-backend/app/schemas/additional_work.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.base import BaseResponse, CurrencyCodeMixin


class AdditionalWorkBase(BaseModel):
    name: str
    detail: str | None = None
    # Single price column — value lives in the currency of the FK. The
    # wire format accepts the currency code ('ARS' / 'USD') and the
    # service translates it into `currency_id` on save.
    price: float = 0.0
    currency: str = "ARS"
    # Pricing mode. `flat` (default) uses `price * quantity`. `frente`
    # applies the formula (material.price_m2 * 0.13) + formula_constant,
    # multiplied by linear meters supplied on the budget row.
    type: str = "flat"
    # Constant used in the `frente` formula (units match the material
    # currency; default 1.15 for newly created `frente` rows).
    formula_constant: float | None = None


class AdditionalWorkCreate(AdditionalWorkBase):
    pass


class AdditionalWorkUpdate(BaseModel):
    name: str | None = None
    detail: str | None = None
    price: float | None = None
    currency: str | None = None
    type: str | None = None
    formula_constant: float | None = None


class AdditionalWorkResponse(AdditionalWorkBase, CurrencyCodeMixin, BaseResponse):
    id: int
    created_at: datetime
    updated_at: datetime

File: afamar-frontend/src/types/additionalWork.ts
/** CRUD item for the `additional_works` catalogue. Wire format mirrors
 *  the backend Pydantic schema (currency is a string code; the service
 *  layer translates to `currency_id` FK on save).
 *
 *  `type` controls the pricing mode at budget save time:
 *    - `'flat'` (default): legacy `price * quantity` math.
 *    - `'frente'`: formula `(material.price_m2 * 0.13) * formula_constant`
 *      applied per linear meter supplied in the budget snapshot. */
export interface AdditionalWork {
  id: number;
  name: string;
  detail: string | null;
  price: number;
  currency: 'ARS' | 'USD';
  /** 'flat' (legacy `price * quantity` math) or 'frente'
   *  (formula `(material.price_m2 * 0.13) * formula_constant`,
   *  multiplied by `linear_meters` from the budget row). */
  type: 'flat' | 'frente';
  /** Multiplicador (default 1.15). Solo aplica a `frente`. DB column name
   *  sigue siendo `formula_constant` (no se renombró para evitar una
   *  migración adicional). */
  formula_constant: number | null;
  created_at?: string;
  updated_at?: string;
}

/** Catalogue picklist for the picker. */
export const ADDITIONAL_WORK_TYPES = ['flat', 'frente'] as const;
export type AdditionalWorkType = (typeof ADDITIONAL_WORK_TYPES)[number];
